docs/source/clickhouse/schema_design.rst

```diff
@@ -16,31 +18,37 @@ data in a column that contains two ``Nested`` arrays where the first array conta
 your dataset design gives you the ability to filter based on other attributes to a small
 number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
-Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
-specific ``key=value`` pair. In this case one is often best-served by creating a top-level
-column and "promoting" the data to it [#dupe]_.
-
-This promotion is preferable on a couple dimensions:
-
-1. Even though it looks like a dictionary, our tags and contexts "maps" are actually arrays of keys and values.
-   With array accesses Clickhouse needs to locate the index of the key whose value needs to be looked up.
-   That is an O(N) lookup. The same key may be located at different indexes for each row.
-   Once the index is known, looking up its value is O(1) since you perform direct index access.
-   If the query has to lookup M rows, this becomes an O(M * N) in terms of time complexity.
-
-   With direct column access, there is no index lookup for a key. The overall complexity to lookup M rows
-   for "promoted columns" is O(M).
-2. Queries that filter based on an array column may have to load the entire column into memory
-   in order to do filtering. Adding a new index on top of a top-level column is a more
-   straightforward operation that guarantees more consistent performance outcomes vs. iterating
-   over an array
+Often, however, a ClickHouse user is predominantly interested in rows that contain a specific tag key or a
+specific ``key=value`` pair that occurs in a very large percentage of the table data (for example,
+``http_status_code=200`` or ``hasKey(http_status_code)``). This breaks any efficiency
+of the bloom-filter style index on dictionary columns (see :ref:`bloom`).
+In this case one is often best-served by creating a top-level column and "promoting"
+the data to it [#dupe]_.
+
+It is very efficient for queries to filter by a specific matching value that exists in
+a high number of rows for a given tag key being non-empty or key:value pair on a column
+that contains **just** the values of that specific key. If we know the tag key is low-selectivity,
+we can get a significant performance improvement with a relatively straightforward schema
+modification. Note that this does require creating an index on the new promoted column.
+
+.. _bloom:
+
+Bloom filter indexing on dictionary-like columns
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+To facilitate faster searching on dictionary columns, we tend to create bloom filter indices
+on a hash of both the unique ``key`` values of each row as well as the ``key=value`` pairs
+of each row. The `bloom filter <https://en.wikipedia.org/wiki/Bloom_filter>`_ index that stores these
+is a stochastic data structure designed to quickly determine which elements do NOT exist in a set,
+but provides little to no optimization if a value is often present in the underlying set.
```

Comment:
In this case it actually tend to be counter productive as you have to load the index first., you do not drop anything, and then you have to scan the whole column. Without the index you only scan the column.

Reply:
good point, changing this wording